### PR TITLE
rebuild index for new prop

### PR DESCRIPTION
**Pull request by sworduo**
#### What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### What does this PR do?
rebuild tag index with old schema version value

#### Which issue(s)/PR(s) this PR relates to?
https://github.com/vesoft-inc/nebula/issues/3274#ref-issue-1046564795
  
#### Checklist：
- [ ] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

**Comment by critical27** [comment]
Thx for contribution, does this PR origin from https://discuss.nebula-graph.com.cn/t/topic/6376?

**Comment by sworduo** [comment]
> Thx for contribution, does this PR origin from https://discuss.nebula-graph.com.cn/t/topic/6376?

Yes

**Review by critical27 — COMMENTED** [review]
I have a suggestion to fix this issue:
1. Perhaps don't modify the `RowReader` or `RowReaderWrapper`, because `RowReader` is bound to a specific schema by design, and the `setLatestSchema` method will introduce meny problem. (e.g. read from an old schema by latest schema will have undefined behaviors)
2. To fix this issue, when we call `reader->getValueByName` in `IndexKeyUtils::collectIndexValues`, we will return `NullType::UNKNOWN_PROP` for a not existed property in old schema (see line 55 in `RowReaderV2.cpp`). When you find the value is unknown, we can use latest schema to check if it is nullable or has default value:
    * return error when it isn't nullable and does not have default value, or skip it just like you did in the PR
    * or use the predefined value to build the index

**Comment by sworduo** [comment]
> I have a suggestion to fix this issue:
> 
> 1. Perhaps don't modify the `RowReader` or `RowReaderWrapper`, because `RowReader` is bound to a specific schema by design, and the `setLatestSchema` method will introduce meny problem. (e.g. read from an old schema by latest schema will have undefined behaviors)
> 2. To fix this issue, when we call `reader->getValueByName` in `IndexKeyUtils::collectIndexValues`, we will return `NullType::UNKNOWN_PROP` for a not existed property in old schema (see line 55 in `RowReaderV2.cpp`). When you find the value is unknown, we can use latest schema to check if it is nullable or has default value:
>    
>    * return error when it isn't nullable and does not have default value
>    * or use the predefined value to build the index

1.The scheme used to read data is the schema included in the row data instead of the latest schema. The latest schema is only used in **IndexKeyUtils::collectIndexValues** when the schema included in the data is different from the latest schema. Besides, since the index will be deleted when the associated prop is deleted,  I'm not very sure what problem will be introduced.

2.In this case, a new parameters named latestSchema need to be added in **IndexKeyUtils::collectIndexValues** and everywhere call this function need to modify as well. Of course, that's OK.

**Comment by sworduo** [comment]
By the way, I wonder how to call pre-commit by my-self before commit...

**Comment by critical27** [comment]
> Details

You mean the lint failed? We need to do the `git clang-format`

**Comment by critical27** [comment]
> 1.The scheme used to read data is the schema included in the row data instead of the latest schema. The latest schema is only used in **IndexKeyUtils::collectIndexValues** when the schema included in the data is different from the latest schema. Besides, since the index will be deleted when the associated prop is deleted, I'm not very sure what problem will be introduced.
> 
> 2.In this case, a new parameters named latestSchema need to be added in **IndexKeyUtils::collectIndexValues** and everywhere call this function need to modify as well. Of course, that's OK.

Perhaps I didn't clarify very clearly. The reason I don't suggest to add `setLatestSchema` is we don't need to do it (`RowReader` should not have latest schema). And the for each time `getTagPropReader` and `getEdgePropReader` we need to call `schemaMan->getLatestTagSchemaVersion`, it could raise a performance issue. The latest schema only need to be fetched once.

As for `a new parameters named latestSchema` you said, either use a default nullptr parameter, or add a overloaded version  which has 3 parameters of `IndexKeyUtils::collectIndexValues`

**Comment by sworduo** [comment]
> > 1.The scheme used to read data is the schema included in the row data instead of the latest schema. The latest schema is only used in **IndexKeyUtils::collectIndexValues** when the schema included in the data is different from the latest schema. Besides, since the index will be deleted when the associated prop is deleted, I'm not very sure what problem will be introduced.
> > 2.In this case, a new parameters named latestSchema need to be added in **IndexKeyUtils::collectIndexValues** and everywhere call this function need to modify as well. Of course, that's OK.
> 
> Perhaps I didn't clarify very clearly. The reason I don't suggest to add `setLatestSchema` is we don't need to do it (`RowReader` should not have latest schema). And the for each time `getTagPropReader` and `getEdgePropReader` we need to call `schemaMan->getLatestTagSchemaVersion`, it could raise a performance issue. The latest schema only need to be fetched once.
> 
> As for `a new parameters named latestSchema` you said, either use a default nullptr parameter, or add a overloaded version which has 3 parameters of `IndexKeyUtils::collectIndexValues`. Both ways LGTM.

That's OK. The point is to avoid call getLatestSchem lots of time.

File: src/common/utils/IndexKeyUtils.cpp [hunk 1]
```diff
@@ -156,25 +157,58 @@ Value IndexKeyUtils::parseIndexTTL(const folly::StringPiece& raw) {
...
   return encodeValues(std::move(values), cols);
 }
 
+// static
+StatusOr<Value>
+IndexKeyUtils::readValueWithLatestSche(RowReader* reader,
+                                       const std::string propName,
+                                       const meta::SchemaProviderIf* latestSchema) {
+  auto value = reader->getValueByName(propName);
+  if (latestSchema == nullptr
```
Comment (critical27) [inline src/common/utils/IndexKeyUtils.cpp:193]: Perhaps there is one more condition here to check if the value is a "Null" (literally, sorry about the naming). You only handle the UNKNOWN_PROP.

See the code in RowReader
image

**Review by critical27 — COMMENTED** [review]
Good job, only one thing to check, please see the inline comments.

PS: some lint failed, please fix them according to https://github.com/vesoft-inc/nebula/runs/4319257516?check_suite_focus=true

File: src/common/utils/IndexKeyUtils.cpp (hunk 1 above)
Comment (sworduo) [reply src/common/utils/IndexKeyUtils.cpp:193]: The condition NullType::__NULL__ is handled by `IndexKeyUtils::checkValue`. Maybe I don't need to handle it in function  `readValueWithLatestSche`.

Comment (critical27) [reply src/common/utils/IndexKeyUtils.cpp:193]: I mean if the value in RowReader is the `NullType::NULL`, we should directly return it. What do you think?

Comment (critical27) [reply src/common/utils/IndexKeyUtils.cpp:193]: Ah, my bad, I see the point here. Never mind.

**Review by critical27 — APPROVED** [review]
LGTM now. Thanks a lot.

**Comment by critical27** [comment]
Em.. the code format failed again. You could try to read the [conduct](https://github.com/vesoft-inc/nebula/blob/a14d7b4192b56d0a7359597bdf755a29044e366a/CODE_OF_CONDUCT.md)

**Comment by sworduo** [comment]
> Em.. the code format failed again. You could try to read the [conduct](https://github.com/vesoft-inc/nebula/blob/a14d7b4192b56d0a7359597bdf755a29044e366a/CODE_OF_CONDUCT.md)

Thx, I will try it

**Comment by yixinglu** [comment]
> > Em.. the code format failed again. You could try to read the [conduct](https://github.com/vesoft-inc/nebula/blob/a14d7b4192b56d0a7359597bdf755a29044e366a/CODE_OF_CONDUCT.md)
> 
> 
> 
> Thx, I will try it

hi @sworduo you can lookup the readme in the tests directory to fix the feature format lint error.

**Comment by sworduo** [comment]
> > > Em.. the code format failed again. You could try to read the [conduct](https://github.com/vesoft-inc/nebula/blob/a14d7b4192b56d0a7359597bdf755a29044e366a/CODE_OF_CONDUCT.md)
> > 
> > 
> > Thx, I will try it
> 
> hi @sworduo you can lookup the readme in the tests directory to fix the feature format lint error.

Sorry about the format error. I will fix it tomorrow.

**Comment by sworduo** [comment]
> > > Em.. the code format failed again. You could try to read the [conduct](https://github.com/vesoft-inc/nebula/blob/a14d7b4192b56d0a7359597bdf755a29044e366a/CODE_OF_CONDUCT.md)
> > 
> > 
> > Thx, I will try it
> 
> hi @sworduo you can lookup the readme in the tests directory to fix the feature format lint error.

Hi @yixinglu , when I run make fmt in tests directory. The result is 
```
All done! 💥 💔 💥
136 files reformatted, 3 files left unchanged, 8 files failed to reformat.
```

A lot of files have been reformatted. And the change is like follows : 

```
     When executing query:
-      """
-      CREATE TAG tag1()
-      """
+    """
+    CREATE TAG tag1()
+    """
     Then the execution should be successful
     # if not exists
     When executing query:
-      """
-      CREATE TAG IF NOT EXISTS tag1()
-      """
+    """
+    CREATE TAG IF NOT EXISTS tag1()
+    """
```

However, it is failed to reformat index.feature. The error is : 

```
Error: cannot format /data/luoshangjun/source/nebula/tests/tck/features/index/Index.feature: INTERNAL ERROR: Invalid file contents are produced:
Parser errors:
(939:2): expected: #EOF, #TableRow, #StepLine, #TagLine, #ScenarioLine, #ScenarioOutlineLine, #Comment, #Empty, got '`name`(64)'
(940:1): expected: #EOF, #TableRow, #StepLine, #TagLine, #ScenarioLine, #ScenarioOutlineLine, #Comment, #Empty, got ')" |'
(938:7): inconsistent cell count within the table
(948:2): expected: #EOF, #TableRow, #StepLine, #TagLine, #ScenarioLine, #ScenarioOutlineLine, #Comment, #Empty, got '`age`'
(949:1): expected: #EOF, #TableRow, #StepLine, #TagLine, #ScenarioLine, #ScenarioOutlineLine, #Comment, #Empty, got ')" |'
(947:7): inconsistent cell count within the table
Please report a bug on https://github.com/ducminh-phan/reformat-gherkin/issues.
This invalid output might be helpful:
/tmp/rfmt-ghk_hr7pzs20.log
```

I also run `make check-and-diff -C tests ` which produce the same results. I have no idea how to format it.

**Comment by critical27** [comment]
You probably need to rebase onto latest master, which will works fine.

**Comment by sworduo** [comment]
> You probably need to rebase onto latest master, which will works fine.

I have been rebased the latest master while produce the same results. Fortunately, formatting is successful in docker...

**Comment by cangfengzhs** [comment]
LGTM and please deal with the conflict.

**Comment by sworduo** [comment]
> LGTM and please deal with the conflict.

That's OK. I have been resolved the conflict in my machine. However, I can not upgrade to third-party 3.0 to compile the code. The information is here https://github.com/vesoft-inc/nebula/pull/3462#issuecomment-1001838415

**Comment by Sophie-Xie** [comment]
> > LGTM and please deal with the conflict.
> 
> That's OK. I have been resolved the conflict in my machine. However, I can not upgrade to third-party 3.0 to compile the code. The information is here [#3462 (comment)](https://github.com/vesoft-inc/nebula/pull/3462#issuecomment-1001838415)

Can you push your code, then try to compile it by CI of repo first. 😂

**Review by critical27 — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**